### PR TITLE
PIFR-16 Zomato to redux

### PR DESCRIPTION
**Pull request by Philips7**

### CODE REVIEW

File: src/app/app.component.ts [hunk 1]
```diff
@@ -1,159 +1,26 @@
...
-import {HttpErrorResponse} from '@angular/common/http';
 import {Store} from '@ngrx/store';
 import {Observable} from 'rxjs/Observable';
-import {AngularFireDatabase} from 'angularfire2/database';
 
-import {ICollection, IEstablishment, IGeocode, LocationService,
-  ILocation, ILocationDetails, ICuisines, ICity, User, CommonService} from './features';
+import {GeoCode, User} from './features';
 import {AppState, GetUser, GoogleLogin, Logout} from './store';
+import {GetGeocode} from './store/actions/geocode.actions';
```
Comment (desfero) [inline src/app/app.component.ts]: Why `GetGeocode` action is inside store, maybe we should put it inside some feature module or core module?

File: src/app/app.component.ts [hunk 2]
```diff
@@ -1,159 +1,26 @@
...
-      });
-
+  constructor(private store: Store<AppState>) {
   }
 
-
   ngOnInit(): void {
     this.user$ = this.store.select('user');
-
+    this.geocode$ = this.store.select('geocode');
```
Comment (desfero) [inline src/app/app.component.ts]: Refactor to selectors, maybe lets tommorow discuss this topic. Ping me on skype for this

File: src/app/features/restaurants/interfaces/IGeocodeResponse.ts [hunk 3]
```diff
@@ -8,3 +8,11 @@ export interface IGeocodeResponse {
     ];
 
 }
+
+
+export class GeoCode implements IGeocodeResponse {
+  constructor(public link: null, public location: null,
+              public nearby_restaurants: null, public loading?: false, public error?: string) {
```
Comment (desfero) [inline src/app/features/restaurants/interfaces/IGeocodeResponse.ts]: Why you have `null` types everywhere? Maybe something like ILocation, INearbyRestaurants should be put instead?

File: src/app/shared/components/google-maps/google-maps.component.html [hunk 4]
```diff
@@ -1,15 +1,27 @@
...
+  <div *ngIf="geocode$ | async as geocode">
+
+    <agm-marker *ngFor="let restaurant of geocode.nearby_restaurants"
+                [iconUrl]="'assets/pointer-roast.png'"
+                [latitude]="restaurant.restaurant.location.latitude | toNumber"
+                [longitude]="restaurant.restaurant.location.longitude | toNumber">
+
+      <agm-snazzy-info-window [shadow]="shadow" [padding]="'5px 10px'" [border]="border"
+                              [maxWidth]="200" [closeWhenOthersOpen]="true">
+        <ng-template>
```
Comment (desfero) [inline src/app/shared/components/google-maps/google-maps.component.html]: probably you don't need template `ng-tempalte` here. For better understanding what ng-template is reffer to https://toddmotto.com/angular-ngfor-template-element

File: src/app/shared/components/google-maps/google-maps.component.html [hunk 5]
```diff
@@ -1,15 +1,27 @@
...
+          <h4 style="text-align: center">{{restaurant.restaurant.name}}</h4>
```
Comment (desfero) [inline src/app/shared/components/google-maps/google-maps.component.html]: move styles to css file

File: src/app/shared/components/google-maps/google-maps.component.html [hunk 6]
```diff
@@ -1,15 +1,27 @@
...
+          <hr style="border: solid 1px #919191">
+          <p style="text-align: center">Ocena:
+            <span>
```
Comment (desfero) [inline src/app/shared/components/google-maps/google-maps.component.html]: do we need span and strong at the same time? Probably in that case strong is not valid usage as its purpose is to highlight important text

File: src/app/shared/components/google-maps/google-maps.component.html [hunk 7]
```diff
@@ -1,15 +1,27 @@
...
+    <agm-marker *ngFor="let restaurant of geocode.nearby_restaurants"
+                [iconUrl]="'assets/pointer-roast.png'"
+                [latitude]="restaurant.restaurant.location.latitude | toNumber"
+                [longitude]="restaurant.restaurant.location.longitude | toNumber">
+
+      <agm-snazzy-info-window [shadow]="shadow" [padding]="'5px 10px'" [border]="border"
+                              [maxWidth]="200" [closeWhenOthersOpen]="true">
+        <ng-template>
+          <h4 style="text-align: center">{{restaurant.restaurant.name}}</h4>
+          <hr style="border: solid 1px #919191">
```
Comment (desfero) [inline src/app/shared/components/google-maps/google-maps.component.html]: remove hr tag, and add some styles for for e.g. h4

File: src/app/shared/components/google-maps/google-maps.component.html (hunk 5 above)
Comment (desfero) [inline src/app/shared/components/google-maps/google-maps.component.html]: +1 for usage of h4, not h1 or h2

File: src/app/shared/components/google-maps/google-maps.component.ts [hunk 8]
```diff
@@ -12,15 +15,17 @@ import {MapStyle} from './map.style';
...
+  geocode$: Observable<GeoCode>;
   lat;
   lng;
   border: object;
   shadow: object;
 
-  constructor() {
-    this.setUserLocation();
+  constructor(private store: Store<AppState>) {
+    this.geocode$ = this.store.select('geocode');
```
Comment (desfero) [inline src/app/shared/components/google-maps/google-maps.component.ts]: move to selectors

File: src/app/shared/components/google-maps/google-maps.component.ts [hunk 9]
```diff
@@ -38,6 +43,11 @@ export class GoogleMapsComponent {
       navigator.geolocation.getCurrentPosition(position => {
         this.lat = position.coords.latitude;
         this.lng = position.coords.longitude;
+        const geocodeRequest: IGeocode = {
+          lat: this.lat,
+          lon: this.lng,
+        };
+        this.store.dispatch(new GetGeocode(geocodeRequest));
```
Comment (desfero) [inline src/app/shared/components/google-maps/google-maps.component.ts]: i don't understand the purpose of this code, probably tommorow we need to discuss it

File: src/app/shared/components/google-maps/google-maps.component.ts [hunk 10]
```diff
@@ -38,6 +43,11 @@ export class GoogleMapsComponent {
       navigator.geolocation.getCurrentPosition(position => {
         this.lat = position.coords.latitude;
         this.lng = position.coords.longitude;
+        const geocodeRequest: IGeocode = {
+          lat: this.lat,
+          lon: this.lng,
```
Comment (desfero) [inline src/app/shared/components/google-maps/google-maps.component.ts]: keep consistency, should be `lng` instead of `lon`

File: src/app/shared/pipes/to-number.pipe.ts [hunk 11]
```diff
@@ -0,0 +1,12 @@
+import { Pipe, PipeTransform } from '@angular/core';
+
+@Pipe({
+  name: 'toNumber'
+})
+export class ToNumberPipe implements PipeTransform {
+
+  transform(value: any, args?: any): any {
```
Comment (desfero) [inline src/app/shared/pipes/to-number.pipe.ts]: make correct type annotation, value: string, remove args, and return type is number

File: src/app/shared/pipes/to-number.pipe.ts [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+    return parseFloat(value);
```
Comment (desfero) [inline src/app/shared/pipes/to-number.pipe.ts]: as a side node, you can also parse like this `+value;`

File: src/app/store/actions/geocode.actions.ts [hunk 13]
```diff
@@ -0,0 +1,31 @@
+import {Action} from '@ngrx/store';
+
+export const GET_GEOCODE = '[GEOCODE] Get Geocode';
+export const GET_GEOCODE__SUCCESS = '[GEOCODE] Get Geocode success';
+export const GET_GEOCODE__ERROR = '[GEOCODE] Get Geocode error';
+
+export class GetGeocode implements Action {
+  readonly type = GET_GEOCODE;
+
+  constructor(public payload?: any) {
```
Comment (desfero) [inline src/app/store/actions/geocode.actions.ts]: forget about `any`. Always try to use correct type for better code self documentation

File: src/app/store/actions/geocode.actions.ts [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+  }
+}
+
+export class GeoGeocodeSuccess implements Action {
+  readonly type = GET_GEOCODE__SUCCESS;
+
+  constructor(public payload?: any) {
```
Comment (desfero) [inline src/app/store/actions/geocode.actions.ts]: as above

File: src/app/store/actions/geocode.actions.ts [hunk 15]
```diff
@@ -0,0 +1,31 @@
+import {Action} from '@ngrx/store';
+
+export const GET_GEOCODE = '[GEOCODE] Get Geocode';
+export const GET_GEOCODE__SUCCESS = '[GEOCODE] Get Geocode success';
```
Comment (desfero) [inline src/app/store/actions/geocode.actions.ts]: +1 intresting approach to split SUCCESS/ERROR by `__`

File: src/app/store/actions/geocode.actions.ts [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+  readonly type = GET_GEOCODE__SUCCESS;
+
+  constructor(public payload?: any) {
+  }
+}
+
+export class GetGeocodeError implements Action {
+  readonly type = GET_GEOCODE__ERROR;
+
+  constructor(public payload?: any) {
```
Comment (desfero) [inline src/app/store/actions/geocode.actions.ts]: as above

File: src/app/store/effects/geocode.effects.ts [hunk 17]
```diff
@@ -0,0 +1,39 @@
+import {Actions, Effect} from '@ngrx/effects';
+import {Injectable} from '@angular/core';
+import {Observable} from 'rxjs/Observable';
+import {HttpClient, HttpHeaders, HttpParams} from '@angular/common/http';
+
+import * as geocodeActions from '../actions';
```
Comment (desfero) [inline src/app/store/effects/geocode.effects.ts]: also forget about `import *`, try to use `import { ... }

File: src/app/store/effects/geocode.effects.ts [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+@Injectable()
+export class GeocodeEffects {
+  headerKey = 'user-key';
+  apiHeader = new HttpHeaders().set(this.headerKey, environment.zomato.apiKey);
+  geocodePath = '/geocode';
+
+  @Effect()
+  getGeocode: Observable<Action> = this.actions.ofType(geocodeActions.GET_GEOCODE)
+    .map((action: geocodeActions.GetGeocode) => action.payload)
+    .switchMap((payload) => {
```
Comment (desfero) [inline src/app/store/effects/geocode.effects.ts]: you don't need paranthesis in payload, should be `payload => ...`

File: src/app/store/effects/geocode.effects.ts [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+      let params = new HttpParams();
+      params = params.append('lat', `${payload.lat}`);
+      params = params.append('lon', `${payload.lon}`);
+
+      return this.http.get(environment.zomato.url + this.geocodePath,
+        {headers: this.apiHeader, params: params});
```
Comment (desfero) [inline src/app/store/effects/geocode.effects.ts]: `this.apiHeader` can be moved as a `const` to `switchMap` because it's newer used only here. We need to keep as low as possible `this.` references in code as they make code less clear and hard to follow

File: src/app/store/effects/geocode.effects.ts [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+
+    }).map(response => {
```
Comment (desfero) [inline src/app/store/effects/geocode.effects.ts]: for convention map should be on next line

File: src/app/store/effects/geocode.effects.ts [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+      const geocodeResponse = new GeoCode(response['link'],
```
Comment (desfero) [inline src/app/store/effects/geocode.effects.ts]: refactor to `response.link`, `response.location`, `response.nearby_restaurants`

File: src/app/store/effects/geocode.effects.ts [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+        response['location'], response['nearby_restaurants']);
+      return new geocodeActions.GeoGeocodeSuccess(geocodeResponse);
+    })
+    .catch( err => Observable.of(new geocodeActions.GetGeocodeError(({error: err.message}))));
```
Comment (desfero) [inline src/app/store/effects/geocode.effects.ts]: use simple `of` method

Comment (desfero) [inline src/app/store/effects/geocode.effects.ts]: remove space before `err`

File: src/app/store/interfaces/AppState.ts [hunk 23]
```diff
@@ -1,5 +1,6 @@
-import {User} from '../../features/';
+import {User, GeoCode} from '../../features/';
```
Comment (desfero) [inline src/app/store/interfaces/AppState.ts]: remove last `/`

File: src/app/store/reducers/geocode.reducer.ts [hunk 24]
```diff
@@ -0,0 +1,19 @@
+import * as geocodeActions from '../actions/geocode.actions';
```
Comment (desfero) [inline src/app/store/reducers/geocode.reducer.ts]: refactor to `import {...}`

File: src/app/store/reducers/geocode.reducer.ts [hunk 25]
```diff
@@ -0,0 +1,19 @@
...
+export function geocodeReducer(state: GeoCode = defaultGeocodeState, action: Action) {
+  switch (action.type) {
+    case geocodeActions.GET_GEOCODE:
+      return {...state, ...action.payload, loading: true};
+    case geocodeActions.GET_GEOCODE__SUCCESS:
+      return {...state, ...action.payload, loading: false};
+    case geocodeActions.GET_GEOCODE__ERROR:
+      return {...state, ...action.payload, loading: false};
+    default:
+      return {...state};
```
Comment (desfero) [inline src/app/store/reducers/geocode.reducer.ts]: you can simly `return state;`

File: src/app/app.component.ts (hunk 1 above)
Comment (Philips7) [reply src/app/app.component.ts]: So if actions shoudn't be placed under action dictionary in a store module, their place is in a core? What about effects and reducers?

File: src/app/shared/components/google-maps/google-maps.component.html (hunk 5 above)
Comment (Philips7) [reply src/app/shared/components/google-maps/google-maps.component.html]: Styles were moved in a next task, I coudn't do it here because of .css extension which I've changed to .sccs

File: src/app/shared/components/google-maps/google-maps.component.html (hunk 7 above)
Comment (Philips7) [reply src/app/shared/components/google-maps/google-maps.component.html]: Removed. I left todo task not to forget it. Is it ok to make use of the ::after selector for this case?

File: src/app/shared/components/google-maps/google-maps.component.html (hunk 4 above)
Comment (Philips7) [reply src/app/shared/components/google-maps/google-maps.component.html]: According to Angular Google Maps docs ng-template is necessary.

Comment (desfero) [reply src/app/shared/components/google-maps/google-maps.component.html]: Yeah, you are right, they query by `TemplateRef` (`@ContentChild(TemplateRef) _templateRef:TemplateRef<any>;`
